Add demo documentation and workflow diagrams (#1224)
aws/aws-iot-device-sdk-embedded-C: docs/doxygen/demos/demos_subpages.txt

/**
@page mqtt_demo_basic_tls Basic TLS Demo

> @brief Uses OpenSSL to establish TLS connection with server authentication, and demonstrates QoS 2 implementation.

Reviewer: ```suggestion
@brief Uses OpenSSL to establish a TLS connection with server authentication, and demonstrates QoS 2 implementation.
```

> @brief Uses OpenSSL to establish a TLS connection with server authentication, and demonstrates QoS 2 implementation.

Reviewer 2: ```suggestion
@brief Uses OpenSSL based transport interface implementation to establish a TLS connection with server authentication, and demonstrates QoS 2 implementation.
```

> This demo demonstrates the subscribe-publish workflow of MQTT. After subscribing to a single topic filter, it publishes to the same topic and waits for receipt of that message to be returned from the server. Once the incoming message is received, the demo publishes an acknowledgement message back to the MQTT server, and the QoS 2 workflow is continued and completed to guarantee the delivery of a message.

Reviewer 2: ```suggestion
This demo demonstrates the subscribe-publish workflow of MQTT. After subscribing to a single topic filter, it publishes to the same topic and waits for receipt of that message back from the server. Once the incoming message is received, the demo publishes an acknowledgement message back to the MQTT server, and the QoS 2 workflow is continued and completed to guarantee the delivery of a message.
```

Reviewer 2: Similarly, for the other demo descriptions

Messages in this demo are sent at QoS 2, which guarantees exactly one delivery according to the MQTT spec. See the demo workflow below:

> @image html basic_tls.png "Basic TLS Demo Workflow" width=90%

Reviewer: Did you use drawio to create these? Could you export them using **File > Export as > PNG...** so that there is not a graph in the background?

*/

/**
@page mqtt_demo_mutual_auth Mutual Authentication Demo

> @brief Uses OpenSSL to establish a TLS connection with server and client authentication, and demonstrates QoS 1 implementation.

Reviewer 2: ```suggestion
@brief Uses OpenSSL based transport interface implementation to establish a TLS connection with server and client authentication, and demonstrates QoS 1 implementation.
```

> @brief Uses OpenSSL to establish TLS connection with server and client authentication, and demonstrates QoS 1 implementation.

Reviewer: ```suggestion
@brief Uses OpenSSL to establish a TLS connection with server and client authentication, and demonstrates QoS 1 implementation.
```

This demo demonstrates the subscribe-publish workflow of MQTT. After subscribing to a single topic filter, it publishes to the same topic and waits for receipt of that message to be returned from the server. Once the incoming message is received, the demo publishes an acknowledgement message back to the MQTT server, and the QoS 1 workflow is completed to guarantee at least one delivery of a message.

Messages in this demo are sent at QoS 1, which guarantees at least one delivery according to the MQTT spec. See the demo workflow below:

@image html mutual_auth.png width=90%
*/

/**
@page mqtt_demo_plaintext Plaintext Demo

> @brief Uses plaintext sockets to establish a TCP connection, and demonstrates QoS 0 implementation.

Reviewer 2: ```suggestion
@brief Uses POSIX socket based transport interface implementation to establish a TCP connection, and demonstrates QoS 0 implementation.
```

> @brief Uses plaintext sockets to establish TCP connection, and demonstrates QoS 0 implementation.

Reviewer: ```suggestion
@brief Uses plaintext sockets to establish a TCP connection, and demonstrates QoS 0 implementation.
```

This demo demonstrates the subscribe-publish workflow of MQTT. After subscribing to a single topic filter, it publishes to the same topic and waits for receipt of that message to be returned from the server.

Messages in this demo are sent at QoS 0, which guarantees at most one delivery according to the MQTT spec. See the demo workflow below:

@image html plaintext.png width=90%
*/

/**
@page mqtt_demo_serializer Serializer Demo

> @brief Uses plaintext sockets to establish a TCP connection, and demonstrates QoS 0 implementation using lightweight serializer API.

Reviewer 2: ```suggestion
@brief Uses POSIX sockets to establish a TCP connection, and demonstrates QoS 0 implementation using lightweight serializer API.
```

> @brief Uses plaintext sockets to establish TCP connection, and demonstrates QoS 0 implementation using lightweight serializer API.

Reviewer: ```suggestion
@brief Uses plaintext sockets to establish a TCP connection, and demonstrates QoS 0 implementation using lightweight serializer API.
```

/**
@page mqtt_demo_basic_tls Basic TLS Demo

> @brief Uses an OpenSSL-based transport interface implementation to establish a TLS connection with server authentication, and demonstrates QoS 2 implementation.

Reviewer 2: Suggest keeping the brief small and concise. The detail of the OpenSSL based transport interface implementation can be mentioned in the paragraph.
```suggestion
@brief Demo of an MQTT application that uses QoS 2 level of communication with broker over a TLS connection for server and client authentication.```

> @brief Demo of a MQTT application that establishes a TLS connection with server-only authentication, and uses QoS 2 level of communication with broker.

Reviewer 2: (Nit) `a MQTT` -> `an MQTT` at all places
```suggestion
@brief Demo of an MQTT application that establishes a TLS connection with server-only authentication, and uses QoS 2 level of communication with broker.
```

/**
@page mqtt_demo_mutual_auth Mutual Authentication Demo

> @brief Uses an OpenSSL--based transport interface implementation to establish a TLS connection with server and client authentication, and demonstrates QoS 1 implementation.

Reviewer 2: ```suggestion
@brief Demo of an MQTT application that uses QoS 1 level of communication with broker over a TLS connection with server and client authentication.
```

/**
@page mqtt_demo_plaintext Plaintext Demo

> @brief Uses a POSIX socket based transport interface implementation to establish a TCP connection, and demonstrates QoS 0 implementation.

Reviewer 2: Please mention about use of POSIX sockets based transport interface implementation in the paragraph

```suggestion
@brief Demo of an MQTT application that uses QoS 0 level of communication with broker over a plain TCP connection with server.
```

/**
@page mqtt_demo_serializer Serializer Demo

> @brief Uses a POSIX socket based transport interface implementation to establish a TCP connection, and demonstrates QoS 0 implementation using only the MQTT serializer API.

Reviewer 2: This should not mention about the transport implementation (as it does not use the MQTT API that depends on the transport interface)
```suggestion
@brief Uses a POSIX socket API to establish a TCP connection, and demonstrates QoS 0 implementation using only the MQTT serializer API.
```

This demo demonstrates the subscribe-publish workflow of MQTT at QoS 0 level. After subscribing to a single topic filter, it publishes to the same topic and waits for receipt of that message to be returned from the server at QoS 0 level. It demonstrates use of the serializer API as a lightweight alternative to the standard MQTT API. This cycle of publishing to the broker and receiving the same message back from the broker is repeated indefinitely.

Messages in this demo are sent at QoS 0, which guarantees at most one delivery according to the MQTT spec. See the demo workflow below:

@image html serializer.png width=90%
*/

/**
@page mqtt_demo_subscription_manager Susbcription Manager Demo

> @brief Uses an OpenSSL-based transport interface implementation to establish a TLS connection, and demonstrates QoS 1 implementation using a subscription manager to manage multiple subscriptions, register callbacks, and handle wildcard topics.

Reviewer 2: Suggest refactoring the brief to focus on the main purpose of the demo.
```suggestion
@brief Demonstrates an MQTT application that subscribes to multiple topic filters using a subscription manager to manage multiple subscriptions, register different callback for each subscription, and handle wildcard topics. It uses an OpenSSL-based transport interface implementation to establish a TLS connection with the broker, and QoS 1 level of communication in the MQTT connection with the broker.
```

> This demo demonstrates the subscribe-publish workflow of MQTT at QoS 1 level, using a subscription manager. After subscribing to a topic filter, it publishes to the same topic and waits for receipt of that message to be returned from the server at QoS 1 level. It repeats this process for 3 different topic filters, including one wildcard filter. This cycle of publishing to the broker and receiving the same message back from the broker is repeated indefinitely.

Reviewer 2: The text should also mention that the application registers different callbacks with the subscription manager for each topic filter that the application subscribes to. Also, mention that the subscription manager invokes the callback registered for topic filter that matches the topic of an incoming PUBLISH message.
